Reproducible epub builds

michael-berg opened the pull request:
Resolves #867 for all epub variants.

All epub builds (standard, advanced, and kobo/kepub) are now reproducible.
Kindle azw3 builds are not yet reproducible since the Calibre epub->azw3 conversion is not reproducible.

Example reproducible build from the following command sequence:
```
git clone https://github.com/standardebooks/edgar-allan-poe_poetry.git

se build --kobo edgar-allan-poe_poetry
mv edgar-allan-poe_poetry_advanced.epub edgar-allan-poe_poetry_advanced_build1.epub
mv edgar-allan-poe_poetry.epub          edgar-allan-poe_poetry_build1.epub
mv edgar-allan-poe_poetry.kepub.epub    edgar-allan-poe_poetry_build1.kepub.epub

sleep 5

se build --kobo edgar-allan-poe_poetry
mv edgar-allan-poe_poetry_advanced.epub edgar-allan-poe_poetry_advanced_build2.epub
mv edgar-allan-poe_poetry.epub          edgar-allan-poe_poetry_build2.epub
mv edgar-allan-poe_poetry.kepub.epub    edgar-allan-poe_poetry_build2.kepub.epub

sha256sum edgar-allan-poe_poetry_advanced_build?.epub
echo ""
sha256sum edgar-allan-poe_poetry_build?.epub
echo ""
sha256sum edgar-allan-poe_poetry_build?.kepub.epub
```
produces the following output, showing the reproducible epub builds:
```
8c55426baa52bf22cee7b98a8f69819a1783eb4efd07af9c6fb42e19cee52e5d  edgar-allan-poe_poetry_advanced_build1.epub
8c55426baa52bf22cee7b98a8f69819a1783eb4efd07af9c6fb42e19cee52e5d  edgar-allan-poe_poetry_advanced_build2.epub

214d943e14a5304b0507cb33716656cad568fd387e4f719f075fd3195439a84f  edgar-allan-poe_poetry_build1.epub
214d943e14a5304b0507cb33716656cad568fd387e4f719f075fd3195439a84f  edgar-allan-poe_poetry_build2.epub

c12ceb0f0ae9aa94c571a7a275ab37558c7a599d16cd351db5cf0c898fc4089a  edgar-allan-poe_poetry_build1.kepub.epub
c12ceb0f0ae9aa94c571a7a275ab37558c7a599d16cd351db5cf0c898fc4089a  edgar-allan-poe_poetry_build2.kepub.epub
```

If no git history is available, `ReproducibleZipFile` date-time still default to the zip file "zero time" of 1980-01-01.
However, in the normal case of there being a git `last_commit`, the `SOURCE_DATE_EPOCH` environment variable gets set during the build so that `ReproducibleZipFile` will use the date-time of that commit.

The following example demonstrates the latest git timestamp (`2025-08-18 16:05:00 -0500` aka `2025-08-18 21:05:00 UTC`) being used in the `ReproducibleZipFile` for the epub.

```
cd edgar-allan-poe_poetry
git log -1
commit 701ecf552ddf10681fdefb4434147104c1dc0d91 (HEAD -> master, origin/master, origin/HEAD)
Author: Alex Cabal <alex@standardebooks.org>
Date:   2025-08-18 16:05:00 -0500

    Update metadata boilerplate
```

```
unzip -l edgar-allan-poe_poetry_build1.epub
Archive:  edgar-allan-poe_poetry_build1.epub
  Length      Date    Time    Name
---------  ---------- -----   ----
       20  08-18-2025 21:05   mimetype
      247  08-18-2025 21:05   META-INF/container.xml
        0  08-18-2025 21:05   META-INF/
        0  08-18-2025 21:05   epub/
    11581  08-18-2025 21:05   epub/content.opf
        0  08-18-2025 21:05   epub/css/
     6347  08-18-2025 21:05   epub/css/core.css
     2086  08-18-2025 21:05   epub/css/local.css
     2027  08-18-2025 21:05   epub/css/se.css
        0  08-18-2025 21:05   epub/images/
   552983  08-18-2025 21:05   epub/images/cover.jpg
    21867  08-18-2025 21:05   epub/images/logo.png
    33169  08-18-2025 21:05   epub/images/titlepage.png
     7854  08-18-2025 21:05   epub/onix.xml
        0  08-18-2025 21:05   epub/text/
     3038  08-18-2025 21:05   epub/text/colophon.xhtml
    39542  08-18-2025 21:05   epub/text/endnotes.xhtml
     2398  08-18-2025 21:05   epub/text/imprint.xhtml
   179879  08-18-2025 21:05   epub/text/poetry.xhtml
      806  08-18-2025 21:05   epub/text/titlepage.xhtml
     3009  08-18-2025 21:05   epub/text/uncopyright.xhtml
    14498  08-18-2025 21:05   epub/toc.ncx
     8381  08-18-2025 21:05   epub/toc.xhtml
---------                     -------
   889732                     23 files
```

acabal commented:
Great work!

We only use Kindleunpack to extract Kindle files in `se extract-ebook`. It's not used during build. `se build` generates Kindle files using a call to the Calibre CLI.

With that in mind, can you remove the changes to Kindleunpack?

I see that `repro-zipfile`simply sets modified dates to `0` which doesn't seem like a great choice. We can use the `SOURCE_DATE_EPOCH` environmental variable to set the modified date it uses. I suggest setting that to the timestamp of the last git commit of the ebook. This can be accessed via `SeEpub.last_commit` which is either a `GitCommit` object or `None`. You can set the variable using `os.environ["SOURCE_DATE_EPOCH"] = ...`. This can be set somewhere at the top of the `SeEpub.build()` function, maybe around `se_epub_build.py` line 132-ish.

acabal commented:
Suggesting reproducible Kindle conversions would be a good suggestion upstream to Calibre. In my experience the maintainer is open to suggestions.

michael-berg commented:
> Suggesting reproducible Kindle conversions would be a good suggestion upstream to Calibre. In my experience the maintainer is open to suggestions.

Once we get this reproducible epubs PR sorted out here, I'll take a look at the Calibre code and see if I can identify where I can submit a similar PR for Calibre.

acabal commented:
If you're very motivated and are able to understand the Calibre epub -> azw3 pipeline, one long-standing wishlist item is to have our own code generate azw3s, and remove the Calibre dependency. We have to install all of Calibre but only use that one function so it would be very nice to not have to do that.

michael-berg commented:
> If you're very motivated and are able to understand the Calibre epub -> azw3 pipeline, one long-standing wishlist item is to have our own code generate azw3s, and remove the Calibre dependency. We have to install all of Calibre but only use that one function so it would be very nice to not have to do that.

I saw that on the list. I'll keep it in mind while I'm looking at the code. No promises, I haven't looked at any of the Calibre code yet...

michael-berg commented:
@acabal For preferred workflows on standardebooks tools, do you prefer that I make the requested changes left as distinct commits in the commit chain on this PR? Or do you prefer that I squash the cumulative set of changes down into a single commit? I know various projects sometimes have strong feelings one way or the other.

michael-berg commented:
@acabal I have implemented your feedback, squashed the changes into a single commit for this PR, and updated the PR title and description to match the new cumulative changes.

acabal commented on se/se_epub_build.py:
```diff
@@ -195,6 +195,7 @@ def build(self, run_epubcheck: bool, check_only: bool, build_kobo: bool, build_k
 					file.truncate()
 
 		# Update the release date in the metadata and colophon.
+		last_updated = None
```
I don't see this getting set to a commit timestamp anywhere, won't it just always be `None` in practice?
